crates/uv-interpreter/src/discovery.rs

```diff
@@ -0,0 +1,1113 @@
+    /// An environment was discovered e.g. via `.venv`
+    DiscoveredEnvironment,
+    /// An executable was found in the search path i.e. `PATH`
+    SearchPath,
+    /// An executable was found via the `py` launcher
+    PyLauncher,
+    /// The interpreter was found in the uv toolchain directory
+    ManagedToolchain,
+    // TODO(zanieb): Add support for fetching the interpreter from a remote source
+    // TODO(zanieb): Add variant for: The interpreter path was inherited from the parent process
+}
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    IO(#[from] io::Error),
+
+    /// An error was encountering when retrieving interpreter information.
+    #[error(transparent)]
+    Query(#[from] crate::interpreter::Error),
+
+    /// An error was encountered when interacting with a managed toolchain.
+    #[error(transparent)]
+    ManagedToolchain(#[from] crate::managed::Error),
+
+    /// An error was encountered when inspecting a virtual environment.
+    #[error(transparent)]
+    VirtualEnv(#[from] crate::virtualenv::Error),
+
+    /// An error was encountered when using the `py` launcher on Windows.
+    #[error(transparent)]
+    PyLauncher(#[from] crate::py_launcher::Error),
+
+    #[error("Interpreter discovery for `{0}` requires `{1}` but it is not selected")]
+    SourceNotSelected(InterpreterRequest, InterpreterSource),
+}
+
+/// Lazily iterate over all discoverable Python executables.
+///
+/// In order, we look in:
+///
+/// - The active environment
+/// - A discovered environment (e.g. `.venv`)
+/// - Installed managed toolchains
+/// - The search path (i.e. PATH)
+/// - `py` launcher output
+///
+/// Each location is only queried if the previous location is exhausted.
+/// Locations may be omitted using `sources`, sources that are not selected will not be queried.
+///
+/// If a [`VersionRequest`] is provided, we will skip executables that we know do not satisfy the request
+/// and (as discussed in [`python_executables_from_search_path`]) additional version specific executables may
+/// be included. However, the caller MUST query the returned executables to ensure they satisfy the request;
+/// this function does not guarantee that the executables provide any particular version. See
+/// [`find_interpreter`] instead.
+fn python_executables<'a>(
+    version: Option<&'a VersionRequest>,
+    sources: &'a SourceSelector,
+) -> impl Iterator<Item = Result<(InterpreterSource, PathBuf), Error>> + 'a {
+    // Note we are careful to ensure the iterator chain is lazy to avoid unnecessary work
+    let iter =
+        // (1) The active environment
+        sources.contains(InterpreterSource::ActiveEnvironment).then(||
+            virtualenv_from_env()
+            .into_iter()
+            .map(virtualenv_python_executable)
+            .map(|path| Ok((InterpreterSource::ActiveEnvironment, path)))
+        ).into_iter().flatten()
+        // (2) A discovered environment
+        .chain(
```

Comment:
Really missing `yield` in rust here, chaining different iterators with results is so much nesting.

An alternative to chaining is boxing, not sure if it's better:

```rust
    let mut iter: Box<dyn Iterator<Item = _>> = Box::new(iter::empty());
    // (1) The active environment
    if sources.contains(InterpreterSource::ActiveEnvironment) {
        let active_venv = virtualenv_from_env()
            .into_iter()
            .map(virtualenv_python_executable)
            .map(|path| Ok((InterpreterSource::ActiveEnvironment, path)));
        iter = Box::new(iter.chain(active_venv));
    }
    // (2) A discovered environment
    if sources.contains(InterpreterSource::DiscoveredEnvironment) {
        let discovered_venv = virtualenv_from_working_dir()
            .map(|path| {
                path.map(virtualenv_python_executable)
                    .map(|path| (InterpreterSource::DiscoveredEnvironment, path))
                    .into_iter()
            })
            .map_err(Error::from);
        iter = Box::new(iter.chain(iter::once(discovered_venv).flatten_ok()));
    }
```